Refactor routing to use a dedicated router file

aig787 opened the pull request:
Move functionality from `.tuono/main.rs` to `.tuono/router.rs` to allow users to override `main.rs`. Updated all related references, logic, and tests to use the new router file for route handling. Introduced a modular approach to improve maintainability and clarity in routing logic.

### Checklist

- [x] I have read [Contributing > Pull requests](https://tuono.dev/documentation/contributing/pull-requests)

### Related issue

Fixes #762 

### Overview

Allow the user to override `main.rs` by moving core functionality into a new `.tuono/router.rs` file. If no user supplied `main.rs` file is found, one is created at `.tuono/main.rs` with the content

```
// File automatically generated
// Do not manually change it

mod router;

use tuono_lib::{Mode, Server, axum::Router, tokio, tuono_internal_init_v8_platform};

#[tokio::main]
async fn main() {
    let server = router::init_server().await;
    server.start().await
}
```

and the user can override this by creating a `src/main.rs` with the contents
```
#[path = "../.tuono/router.rs"]
mod router;

use tuono_lib::tokio;

#[tokio::main]
async fn main() {
    let server = router::init_server().await;
    // custom logic here
}
```

and updating `Cargo.toml` to use the overridden main file
```
[[bin]]
name = "tuono"
path = "src/main.rs"
```

Valerioageno reviewed (commented):
Thanks for it! I pretty much like the approach. Please fix the broken pipelines

aig787 commented:
Looks like it was just extra imports - `pnpm test:e2e` passes for me locally now

Valerioageno commented on crates/tuono/tests/cli_build.rs:
What about adding an integration test to assert that when the users define the `src/main.rs` file, the `AXUM_ENTRY_POINT` doesn't?

Valerioageno reviewed (commented):
Overall, looks good to me. Thanks

aig787 commented:
Just pushed another change that exposes the Router so that you can customize which layers are added. I realized that wasn't possible in the previous implementation.

Valerioageno commented on crates/tuono/tests/cli_build.rs:
```diff
@@ -216,3 +221,20 @@ fn build_fails_with_no_config() {
+        .success();
+
+    let temp_tuono_main_rs_path = temp_tuono_project.path().join(".tuono/main.rs");
+    assert!(!temp_tuono_main_rs_path.exists());
```
I'd just add an assertion that the `.tuono/router.rs` file exists to be sure that something got created

Valerioageno commented on crates/tuono/src/app.rs:
```diff
@@ -40,6 +40,7 @@ pub struct App {
     pub route_map: HashMap<String, Route>,
     pub base_path: PathBuf,
     pub has_app_state: bool,
+    pub has_main: bool,
```
Nitpicking, but I would call this `has_custom_main_file`

Valerioageno commented on crates/tuono/src/app.rs:
```diff
@@ -51,14 +52,23 @@ fn has_app_state(base_path: PathBuf) -> std::io::Result<bool> {
     Ok(contents.contains("pub fn main"))
 }
 
+fn has_main(base_path: PathBuf) -> std::io::Result<bool> {
+    let file = File::open(base_path.join("src/main.rs"))?;
+    let mut buf_reader = BufReader::new(file);
+    let mut contents = String::new();
+    buf_reader.read_to_string(&mut contents)?;
+    Ok(contents.contains("async fn main"))
+}
+
```
```suggestion
fn has_custom_main_file(base_path: PathBuf) -> std::io::Result<bool> {
    let file: String = fs::read_to_string("src/main.rs")?;
    Ok(file.contains("async fn main"))
}
```

Valerioageno commented on crates/tuono/src/source_builder.rs:
```diff
@@ -106,42 +114,57 @@ impl SourceBuilder {
     }
 
-    fn generate_axum_source(&self) -> String {
+    fn generate_axum_source(&self) -> (Option<String>, String) {
```
I'd a top level comment to describe what is the returned tuple

```suggestion
    /// Returns a tuple containing:
    /// - The optional `.tuono/main.rs` file content
    /// - The dynamically generated `.tuono/router.rs` file content
    fn generate_axum_source(&self) -> (Option<String>, String) {
```

Valerioageno commented on crates/tuono/src/source_builder.rs:
```diff
@@ -106,42 +114,57 @@ impl SourceBuilder {
+                    "|router| async { router }"
+                },
+            )
+            .replace(
+                "//GET_ROUTER_SIGNATURE//",
+                if app.has_app_state {
+                    "F: Fn(Router<ApplicationState>, ApplicationState) -> Fut,"
+                } else {
+                    "F: Fn(Router) -> Fut,"
```
Nice one! I'd add a test case for asserting this!

Valerioageno commented on crates/tuono/templates/router.rs:
```diff
@@ -0,0 +1,44 @@
+    tuono_internal_init_v8_platform();
+    let router = get_router(f).await;
+    if MODE == Mode::Prod {
+        println!("\n  ⚡ Tuono v0.19.4");
```
```suggestion
        println!("\n  ⚡ Tuono v//VERSION//");
```

Valerioageno commented on crates/tuono/src/source_builder.rs:
```diff
@@ -106,42 +114,57 @@ impl SourceBuilder {
+                "//APP_STATE_USAGE//",
                 if app.has_app_state {
-                    ".with_state(user_custom_state)"
+                    "|router, user_custom_state| async { router.with_state(user_custom_state) }"
```
I'd also add on this condition a test case

aig787 commented:
@Valerioageno I cleaned up the dead code and rebased, but now see these failures. I don't _think_ they're from me? Could you confirm if they're expected or not?

Valerioageno commented:
I rebased by myself to speed up the merge :)
